File: graphvalidetree.py
class Solution(object):
    def validTree(self, n, edges):
        """
        :type n: int
        :type edges: List[List[int]]
        :rtype: bool
        """
        nums = [-1 for _ in xrange(n)]
        def find(i):
            if nums[i] == -1:
                return i
            return find(nums[i])
        for i in xrange(len(edges)):
            x = find(edges[i][0])
            y = find(edges[i][1])
            
            if x == y:
                return False
            nums[x] = y
            
        return len(edges) == n-1

File: cpp/graphvalidetree.cpp
class Solution {
public:
    bool validTree(int n, vector<pair<int, int>>& edges) {
        vector<int> unionSet(n, -1);
        for (auto &e: edges) {
            int x = find(unionSet, e.first);
            int y = find(unionSet, e.second);
            
            if (x == y) return false;
            unionSet[x] = y;
        }
        return n-1 == edges.size();
    }
    int find(vector<int> &unionSet, int e) {
        if (unionSet[e] == -1) return e;
        return find(unionSet, unionSet[e]);
    }  
};
